cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,234 @@
+
+	ctx context.Context
+}
+
+type upgraderInterface interface {
+	getUpgrader() *upgrader
+	init() error
+	validateAndDeriveParams() error
+	printIntroHelpText() error
+	prepare() (*daisy.Workflow, error)
+	upgrade() (*daisy.Workflow, error)
+	retryUpgrade() (*daisy.Workflow, error)
+	reboot() (*daisy.Workflow, error)
+	cleanup() (*daisy.Workflow, error)
+	rollback() (*daisy.Workflow, error)
+}
+
+// Run runs upgrade workflow.
+func Run(p *InputParams) (*daisy.Workflow, error) {
+	return run(&upgrader{InputParams: p})
+}
+
+func run(u upgraderInterface) (*daisy.Workflow, error) {
+	if err := u.init(); err != nil {
+		return nil, err
+	}
+	if err := u.validateAndDeriveParams(); err != nil {
+		return nil, err
+	}
+	if err := u.printIntroHelpText(); err != nil {
+		return nil, err
+	}
+	return runUpgradeWorkflow(u)
+}
+
+func (u *upgrader) getUpgrader() *upgrader {
+	return u
+}
+
+func (u *upgrader) init() error {
+	log.SetPrefix(logPrefix + " ")
+
+	var err error
+	u.ctx = context.Background()
+	computeClient, err = daisyCompute.NewClient(u.ctx, option.WithCredentialsFile(u.Oauth))
+	mgce = &compute.MetadataGCE{}
+	if err != nil {
+		return daisy.Errf("Failed to create GCE client: %v", err)
+	}
+	return nil
+}
+
+func (u *upgrader) printIntroHelpText() error {
+	guide, err := getIntroHelpText(u)
+	if err != nil {
+		return err
+	}
+	fmt.Print(guide, "\n\n")
+	return nil
+}
+
+func runUpgradeWorkflow(u upgraderInterface) (*daisy.Workflow, error) {
+	var err error
+
+	// If upgrade failed, run cleanup or rollback before exiting.
+	defer func() {
+		handleFailure(u, err)
+	}()
+
+	// step 1: preparation - take snapshot, attach install media, backup/set startup script
+	fmt.Print("\nPreparing for upgrade...\n\n")
+	prepareWf, err := u.prepare()
+	if err != nil {
+		return prepareWf, err
+	}
+
+	// step 2: run upgrade.
+	fmt.Print("\nRunning upgrade...\n\n")
+	upgradeWf, err := u.upgrade()
+	if err == nil {
+		return upgradeWf, nil
+	}
+
+	// step 3: reboot if necessary.
+	if !needReboot(err) {
+		return upgradeWf, err
+	}
+	fmt.Print("\nRebooting...\n\n")
+	rebootWf, err := u.reboot()
+	if err != nil {
+		return rebootWf, err
+	}
+
+	// step 4: retry upgrade.
+	fmt.Print("\nRetrying upgrade...\n\n")
+	retryUpgradeWf, err := u.retryUpgrade()
+	return retryUpgradeWf, err
+}
+
+func handleFailure(ui upgraderInterface, err error) {
+	u := ui.getUpgrader()
+	if err == nil {
+		fmt.Printf("\nSuccessfully upgraded instance '%v' to %v!\n", u.instanceURI, u.TargetOS)
+		// TODO: update the help guide link. b/154838004
+		fmt.Printf("\nPlease verify your applications' functionality of " +
+			"the instance. If it has a problem and can't be fixed, please manually " +
+			"rollback following the guide.\n\n")
+		return
+	}
+
+	isNewOSDiskAttached := isNewOSDiskAttached(u.instanceProject, u.instanceZone, u.instanceName, u.newOSDiskName)
+	if u.AutoRollback {
+		if isNewOSDiskAttached {
+			fmt.Printf("\nFailed to finish upgrading. Rollback to the "+
```

Comment:
the upgrader doing the rolling back. Why using "source OS disk" instead of "original"? My understanding is, there is no "source OS disk", no "source instance". The is only original OS disk and new OS disk. New OS disk is a copy of the original disk, and the upgrade happens in place in the new OS disk. If it success, its OS version will become to "target OS version" from "source OS version". If failed, user needs to detach the new OS disk, attach the original OS disk back. If 'auto rollback' is enabled, it will happen automatically.

With this clarification, do you think the suggested text needs to be modified?

Reply:
Thanks for elaborating. I think the new text can now be: "Upgrade failed to finish. Rolling back to the original state from the original OS disk."